probe/sniff/sniffer.go

```diff
@@ -0,0 +1,247 @@
+	return <-c, nil
+}
+
+func (s *Sniffer) loop(src gopacket.ZeroCopyPacketDataSource, on, off time.Duration) {
+	var (
+		process = uint64(1)               // initially enabled
+		total   = uint64(0)               // total packets seen
+		count   = uint64(0)               // count of packets captured
+		packets = make(chan Packet)       // decoded packets
+		rpt     = report.MakeReport()     // the report we build
+		turnOn  = (<-chan time.Time)(nil) // signal to start capture (initially enabled)
+		turnOff = time.After(on)          // signal to stop capture
+		done    = make(chan struct{})     // when src is finished, we're done too
+	)
+
+	go func() {
+		s.read(src, packets, &process, &total, &count)
+		close(done)
+	}()
+
+	for {
+		select {
+		case p := <-packets:
+			s.Merge(p, rpt)
+
+		case <-turnOn:
+			atomic.StoreUint64(&process, 1) // enable packet capture
+			turnOn = nil                    // disable the on switch
+			turnOff = time.After(on)        // enable the off switch
+
+		case <-turnOff:
+			atomic.StoreUint64(&process, 0) // disable packet capture
+			turnOn = time.After(off)        // enable the on switch
+			turnOff = nil                   // disable the off switch
+
+		case c := <-s.reports:
+			rpt.Sampling.Count = atomic.LoadUint64(&count)
+			rpt.Sampling.Total = atomic.LoadUint64(&total)
+			applySampling(rpt)
+			c <- rpt
+			atomic.StoreUint64(&count, 0)
+			atomic.StoreUint64(&total, 0)
+			rpt = report.MakeReport()
+
+		case <-done:
+			return
+		}
+	}
+}
+
+// applySampling compensates for sampling by artifically inflating counts
+// throughout the report. It should be run once, by the probe, before emitting
+// the report upstream.
+func applySampling(r report.Report) {
+	rate := r.Sampling.Rate()
+	if rate >= 1.0 {
+		return
+	}
+	factor := 1.0 / rate
+	for _, topology := range r.Topologies() {
+		for _, emd := range topology.EdgeMetadatas {
+			if emd.PacketCount != nil {
+				*emd.PacketCount = uint64(float64(*emd.PacketCount) * factor)
+			}
+			if emd.ByteCount != nil {
+				*emd.ByteCount = uint64(float64(*emd.ByteCount) * factor)
+			}
+		}
+	}
+}
+
+// Packet is an intermediate, decoded form of a packet, with the information
+// that the Scope data model cares about. Designed to decouple the packet data
+// source loop, which should be as fast as possible, and the process of
+// merging the packet information to a report, which may take some time and
+// allocations.
+type Packet struct {
+	SrcIP, DstIP       string
+	SrcPort, DstPort   string
+	Network, Transport int // byte counts
+}
+
+func (s *Sniffer) read(src gopacket.ZeroCopyPacketDataSource, dst chan Packet, process, total, count *uint64) {
+	var (
+		p    Packet
+		data []byte
+		err  error
+	)
+	for {
+		data, _, err = src.ZeroCopyReadPacketData()
+		if err == io.EOF {
+			return // done
+		}
+		if err != nil {
+			log.Printf("sniffer: read: %v", err)
+			continue
+		}
+		atomic.AddUint64(total, 1)
+		if atomic.LoadUint64(process) == 0 {
```

Comment:
If I read this correctly, we're always capturing all packets, and only decoding them if we're 'active'?